src/Streams.Core/Streams.fs

```diff
@@ -4,20 +4,18 @@ open System.Collections.Generic
 open System.Threading
 
 /// Provides on-demand iteration 
-type Iterator = {
+type (* internal *) Iterator = 
     /// Function for on-demand processing
-    TryAdvance : unit -> bool 
+    abstract TryAdvance : unit -> bool 
     /// Cleanup function
-    Dispose : unit -> unit 
-}
+    abstract Dispose : unit -> unit 
 
 /// Provides functions for iteration
-type Iterable = {
+type (* internal *) Iterable = 
```

Comment:
OK.  However it's not quite like IEnumerable because we seem to be taking advantage of the fact that we don't call GetEnumerator (= .Iterator property) multiple times - everything is linear. That is, my impression from code like [this](https://github.com/nessos/Streams/blob/master/src/Streams.Core/Streams.fs#L139) is that each Iterable need only ever produce at most one Iterator. The underlying mutable state for the iterator is allocated when the Iterable is created.  We're assuming that the Iterable.Iterator property is only ever called at most once (or if called multiple times it should return a objects sharing unique underlying state)?

If so, then the contract is "For each Iterable, either .Bulk is called once or .Iterator is called once.  If .Iterator is called once, then .Iterator.Dispose is called once".  I think that's workable.

Code like [this](https://github.com/nessos/Streams/blob/master/src/Streams.Core/Streams.fs#L187) then looks problematic because it is run when the Iterable is created. If the iterable's Bulk() method is used instead of .Iterator then the CTS handler will not be deregistered.  It seems we shouldn't be allocating the iterable and hooking it into the CTS until .Iterator is actually needed.  I'll try to see if there's a way to observe this inn a test case.